src/functional/swing_slide_filter.zig

```diff
@@ -0,0 +1,411 @@
+fn appendValue(value: f64, compressed_values: *ArrayList(u8)) !void {
+    const valueAsBytes: [8]u8 = @bitCast(value);
+    try compressed_values.appendSlice(valueAsBytes[0..]);
+}
+
+/// Append `index` to `compressed_values`.
+fn appendIndex(index: usize, compressed_values: *ArrayList(u8)) !void {
+    const indexAsFloat: [8]u8 = @bitCast(@as(f64, @floatFromInt(index)));
+    try compressed_values.appendSlice(indexAsFloat[0..]);
+}
+
+/// Computes the numerator of the slope derivate as in Eq. 6.
+fn computeSlopeDerivate(segment: Segment) f80 {
+    return (segment.end_value - segment.start_value) * usizeToF80(segment.end_time - segment.start_time);
+}
+
+/// Compress `uncompressed_values` within `error_bound` using "Swing Filter"
+/// and write the result to `compressed_values`. If an error occurs it is returned.
+pub fn compress_swing(
+    uncompressed_values: []const f64,
+    compressed_values: *ArrayList(u8),
+    error_bound: f32,
+) Error!void {
+    // Create adjusted error bound to avoid reaching the error bound during decompression.
+    const adjusted_error_bound = if (error_bound > 0)
+        error_bound - tersets.ErrorBoundMargin
+    else
+        error_bound;
+
+    var upper_bound_line: LinearFunction = LinearFunction{ .slope = 0, .intercept = 0 };
+    var lower_bound_line: LinearFunction = LinearFunction{ .slope = 0, .intercept = 0 };
+    var new_upper_bound_line: LinearFunction = LinearFunction{ .slope = 0, .intercept = 0 };
+    var new_lower_bound_line: LinearFunction = LinearFunction{ .slope = 0, .intercept = 0 };
+    var current_line: LinearFunction = LinearFunction{ .slope = 0, .intercept = 0 };
+
+    // Initialize the current segment with first two points.
+    var current_segment: Segment = Segment{
+        .start_time = 0,
+        .start_value = uncompressed_values[0],
+        .end_time = 1,
+        .end_value = uncompressed_values[1],
+    };
+
+    // Compute the numerator Eq. (6).
+    var slope_derivate: f80 = computeSlopeDerivate(current_segment);
+
+    updateSwingLine(current_segment, &upper_bound_line, adjusted_error_bound);
+    updateSwingLine(current_segment, &lower_bound_line, -adjusted_error_bound);
+
+    // First two points already part of `current_segment`, next point is at index 2.
+    var current_timestamp: usize = 2;
+    while (current_timestamp < uncompressed_values.len) : (current_timestamp += 1) {
+        // Evaluate the upper and lower bound lines at current timestamp
+        // to obtain the upper and lower limits of the current point.
+        const upper_limit = evaluateLineAtTime(upper_bound_line, current_timestamp);
+        const lower_limit = evaluateLineAtTime(lower_bound_line, current_timestamp);
+
+        if ((upper_limit < (uncompressed_values[current_timestamp] - adjusted_error_bound)) or
+            (lower_limit > (uncompressed_values[current_timestamp] + adjusted_error_bound)))
+        { // Recording mechanism (the current points is outside the limits).
```

Comment:
Maybe move comments to the next line as it is empty anyway.

Reply:
Fixed.